### PR TITLE
[train] Add fault tolerance variant to the training data ingest benchmark

### PR DESCRIPTION
**Pull request by justinvyu**
## Summary

Adds a `.skip_training.fault_tolerance` variant to the image classification training ingest release test which **kills a node every N seconds and tests the worker recovery.**

Also add the ability to **stress-test concurrent multi-dataset execution with training ingest** (training and validation datasets) by performing validation *during* the training epoch every N steps. This is not enabled because it is too unperformant and will cause the test to run for too long. This will be addressed in a follow-up PR.

Also updates the training benchmark script to load training state properly and do mid-epoch resumption by skipping batches all the way up to the batch that corresponds to the latest checkpoint.

Adds the following metrics:
- `checkpoint/download`: Time spent downloading the checkpoint from storage to local.
- `checkpoint/load`: Time spent loading the checkpoint from local.
- `train/iter_skip_batch`: Time spent skipping batches upon restoration to do "mid-epoch resumption."
- `checkpoint/restoration_time`: Extra time spent on restoration, which is just the sum of the above 3.

Here's what the fault tolerance test does:
- Starts a chaos killer which kills a node every ~480 seconds, killing up to 2 nodes across the entire job.
- Runs training with `max_failures=4` (with 2 extra failures than needed as a buffer).

### CODE REVIEW

File: release/train_tests/benchmark/image_classification/factory.py [hunk 1]
```diff
@@ -51,6 +57,21 @@ def get_ray_datasets(self) -> Dict[str, ray.data.Dataset]:
...
+            train_ds.context.execution_options.exclude_resources = (
+                train_ds.context.execution_options.exclude_resources.add(
+                    ray.data.ExecutionResources(cpu=cpus_to_exclude)
+                )
+            )
+            logger.info(
+                f"[Dataloader] Reserving {cpus_to_exclude} CPUs for validation "
+                "that happens concurrently with training every "
+                f"{self.benchmark_config.validate_every_n_steps} steps. "
+            )
```
Comment (justinvyu) [inline release/train_tests/benchmark/image_classification/factory.py:73]: @raulchen Is this a reasonable way to handle multi-dataset?

File: release/train_tests/benchmark/train_benchmark.py [hunk 2]
```diff
@@ -230,6 +269,17 @@ def get_metrics(self) -> Dict[str, float]:
...
+        # Extra time that each worker spends to restore from checkpoint,
+        # which includes downloading the checkpoint, loading the checkpoint,
+        # and skipping through batches that were already processed.
+        restoration_time = (
+            self._metrics["checkpoint/download"].get()
+            + self._metrics["checkpoint/load"].get()
+            + self._metrics["train/iter_skip_batch"].get()
+        )
+        if restoration_time > 0:
+            metrics["checkpoint/restoration_time"] = restoration_time
```
Comment (justinvyu) [inline release/train_tests/benchmark/train_benchmark.py:281]: Note: We should track "restoration time" in Ray Train by default. This time does not include process group re-init, actor startup, etc. Would be good to sum these two together to get the restoration startup cost.

File: release/train_tests/benchmark/train_benchmark.py [hunk 3]
```diff
@@ -57,22 +70,28 @@ def run(self):
...
         with self._metrics["train/iter_first_batch"].timer():
-            batch = self.get_next_batch(self.train_dataloader)
+            batch = self.get_next_batch(train_dataloader)
 
-        # TODO: Handle the case where we restored to a middle of the epoch.
+        # Skip through batches if we restored to a middle of the epoch.
+        # TODO: Compare this baseline to the data checkpointing approach once we have it.
+        for _ in range(self._train_batch_idx):
+            with self._metrics["train/iter_skip_batch"].timer():
+                self.get_next_batch(train_dataloader)
```
Comment (justinvyu) [inline release/train_tests/benchmark/train_benchmark.py:94]: Keeping an accurate total for the time spent "skipping" batches is a little tricky because this step could take a long time (ex: if training was killed previously at the second to last batch). Then, a node could get killed in this time, and the time spent skipping batches would be lost from the metrics, which are only snapshotted every checkpoint.

May want to checkpoint the metrics separately from the model if we want to track this better.

File: release/train_tests/benchmark/train_benchmark.py [hunk 4]
```diff
@@ -27,25 +31,34 @@ def __init__(self, factory: BenchmarkFactory):
...
+            download_start = time.perf_counter()
+            checkpoint.to_directory(temp_checkpoint_dir)
+            download_time = time.perf_counter() - download_start
+
+            load_start = time.perf_counter()
+            self.load_checkpoint(temp_checkpoint_dir)
+            load_time = time.perf_counter() - load_start
+
+            self._metrics["checkpoint/download"].add(download_time)
+            self._metrics["checkpoint/load"].add(load_time)
```
Comment (matthewdeng) [inline release/train_tests/benchmark/train_benchmark.py:61]: nit: use same context manager pattern as done later with the batch iteration?

e.g.

```python
with self._metrics["checkpoint/download"].timer():
    checkpoint.to_directory(temp_checkpoint_dir)


with self._metrics["checkpoint/load"].timer():
    self.load_checkpoint(temp_checkpoint_dir)
```

File: release/train_tests/benchmark/train_benchmark.py [hunk 5]
```diff
@@ -135,36 +163,51 @@ def validate_and_checkpoint(self):
...
 
             with self._metrics["validation/step"].timer():
-                with torch.no_grad():
-                    out = self.model(input_batch)
-                    loss = self.loss_fn(out, labels)
-                    total_loss += loss
-                    num_rows += len(labels)
-                    self._metrics["validation/rows_processed"].add(len(labels))
+                if not self.benchmark_config.skip_validation_step:
+                    total_loss += self.validate_step(input_batch, labels)
```
Comment (matthewdeng) [inline release/train_tests/benchmark/train_benchmark.py:186]: Inverse ordering of timer and condition? Or do you intentionally want to include this as well like you mentioned in the other comment.

```python
            if not self.benchmark_config.skip_validation_step:
                with self._metrics["validation/step"].timer():
                    total_loss += self.validate_step(input_batch, labels)
```

File: release/train_tests/benchmark/train_benchmark.py [hunk 6]
```diff
@@ -230,6 +293,17 @@ def get_metrics(self) -> Dict[str, float]:
                 metrics["validation/local_throughput"] * num_workers
             )
 
+        # Extra time that each worker spends to restore from checkpoint,
+        # which includes downloading the checkpoint, loading the checkpoint,
+        # and skipping through batches that were already processed.
+        restoration_time = (
+            self._metrics["checkpoint/download"].get()
```
Comment (matthewdeng) [inline release/train_tests/benchmark/train_benchmark.py:300]: nit: Use constants for all keys to make it safe against typos.

**Review by matthewdeng — APPROVED** [review]
Very clean!

File: release/train_tests/benchmark/train_benchmark.py (hunk 4 above)
Comment (justinvyu) [reply release/train_tests/benchmark/train_benchmark.py:61]: `load_checkpoint` will load in the snapshot of the metrics, which will overwrite these guys. So I need to save it separately then add it to the metrics later.

File: release/train_tests/benchmark/train_benchmark.py (hunk 5 above)
Comment (justinvyu) [reply release/train_tests/benchmark/train_benchmark.py:190]: I think it's easier to parse the output if this metric always exists and is just 0 rather than existing conditionally.

File: release/train_tests/benchmark/train_benchmark.py (hunk 6 above)
Comment (justinvyu) [reply release/train_tests/benchmark/train_benchmark.py:306]: I'll do this in a followup.
